python/dgl/nn/pytorch/conv/dgnconv.py

```diff
@@ -0,0 +1,200 @@
+        (torch.sum(
+            torch.abs(eig_s[:, :, eig_idx] - eig_d[:, :, eig_idx]),
+            keepdim=True, dim=1) + 1e-30
+        )
+    ).unsqueeze(-1)
+    h_mod = torch.mul(h, eig_w)
+    return torch.abs(torch.sum(h_mod, dim=1) - torch.sum(eig_w, dim=1) * h_in)
+
+for k in range(1, 4):
+    AGGREGATORS[f'dir{k}-av'] = partial(aggregate_dir_av, eig_idx=k)
+    AGGREGATORS[f'dir{k}-dx'] = partial(aggregate_dir_dx, eig_idx=k)
+
+class DGNConvTower(PNAConvTower):
+    """A single DGN tower with modified reduce function"""
+    def message(self, edges):
+        """message function for DGN layer"""
+        if self.edge_feat_size > 0:
+            f = torch.cat([edges.src['h'], edges.dst['h'], edges.data['a']], dim=-1)
+        else:
+            f = torch.cat([edges.src['h'], edges.dst['h']], dim=-1)
+        return {'msg': self.M(f), 'eig_s': edges.src['eig'], 'eig_d': edges.dst['eig']}
+
+    def reduce_func(self, nodes):
+        """reduce function for DGN layer"""
+        h_in = nodes.data['h']
+        eig_s = nodes.mailbox['eig_s']
+        eig_d = nodes.mailbox['eig_d']
+        msg = nodes.mailbox['msg']
+        degree = msg.size(1)
+
+        h = []
+        for aggregator in self.aggregators:
+            agg_name = aggregator.__name__
+            if agg_name.startswith('dir'):
+                if agg_name.endswith('av'):
+                    h.append(aggregator(msg, eig_s, eig_d))
+                else:
+                    h.append(aggregator(msg, eig_s, eig_d, h_in))
+            else:
+                h.append(aggregator(msg))
+        h = torch.cat(h, dim=1)
+        h = torch.cat([
+            scaler(h, D=degree, delta=self.delta) if scaler is not scale_identity else h
+            for scaler in self.scalers
+        ], dim=1)
+        return {'h_neigh': h}
+
+class DGNConv(PNAConv):
+    r"""Directional Graph Network Layer from `Directional Graph Networks
+    <https://arxiv.org/abs/2010.02863>`__
+
+    The main computation process follows Principal Neighbourhood Aggregation.
+    A PNA layer is composed of multiple PNA towers. Each tower takes as input a split of the
+    input features, and computes the message passing as below.
+
+    .. math::
+        h_i^(l+1) = U(h_i^l, \oplus_{(i,j)\in E}M(h_i^l, e_{i,j}, h_j^l))
+
+    where :math:`h_i` and :math:`e_{i,j}` are node features and edge features, respectively.
+    :math:`M` and :math:`U` are MLPs, taking the concatenation of input for computing
+    output features. :math:`\oplus` represents the combination of various aggregators
+    and scalers. Aggregators aggregate messages from neighbours and scalers scale the
+    aggregated messages in different ways. :math:`\oplus` concatenates the output features
+    of each combination.
+
+    The output of multiple towers are concatenated and fed into a linear mixing layer for the
+    final output.
+    
+    Two additional directional aggregators are supported in DGN according to the vector field
```

Comment:
You can simply say that "DGN introduces two special aggregators." to avoid mentioning PNA.